akka-persistence/src/main/scala/akka/persistence/Persistence.scala

```diff
@@ -73,65 +73,113 @@ final class PersistenceSettings(config: Config) {
+    actorPlugins.put(actor.persistenceId, ActorMemento(actor.journalPluginId, actor.snapshotPluginId))
   }
 
-  private val journal = createPlugin("journal") { clazz ⇒
-    if (classOf[AsyncWriteJournal].isAssignableFrom(clazz)) Dispatchers.DefaultDispatcherId
-    else DefaultPluginDispatcherId
+  /** INTERNAL API. Delete persistent actor memento. */
+  private[persistence] def unregister(actor: Eventsourced): Unit = {
+    actorPlugins.remove(actor.persistenceId)
   }
 
   /**
-   * Creates a canonical persistent actor id from a persistent actor ref.
+   * Returns a journal actor identified by `journalPluginId`.
+   * When empty, looks in `akka.persistence.journal.plugin` to find configuration entry path.
+   * When configured, uses `journalPluginId` as absolute path to the journal configuration entry.
+   * Configuration entry must contain few required fields, such as `class`. See `src/main/resources/reference.conf`.
    */
-  def persistenceId(persistentActor: ActorRef): String = id(persistentActor)
+  def journalBy(journalPluginId: String): ActorRef = {
+    val configPath = if (isDefault(journalPluginId)) config.getString("journal.plugin") else journalPluginId
+    // TODO remove synchronized after https://issues.scala-lang.org/browse/SI-7943
+    journalPlugins.synchronized(journalPlugins.getOrElseUpdate(configPath, createPlugin(configPath)(journalDispatchSelector)))
+  }
 
   /**
-   * Returns a snapshot store for a persistent actor identified by `persistenceId`.
+   * Returns a snapshot store actor identified by `snapshotPluginId`.
+   * When empty, looks in `akka.persistence.snapshot-store.plugin` to find configuration entry path.
+   * When configured, uses `snapshotPluginId` as absolute path to the snapshot store configuration entry.
+   * Configuration entry must contain few required fields, such as `class`. See `src/main/resources/reference.conf`.
    */
-  def snapshotStoreFor(persistenceId: String): ActorRef = {
-    // Currently returns a snapshot store singleton but this methods allows for later
-    // optimizations where each persistent actor can have its own snapshot store actor.
-    snapshotStore
+  def snapshotStoreBy(snapshotPluginId: String): ActorRef = {
+    val configPath = if (isDefault(snapshotPluginId)) config.getString("snapshot-store.plugin") else snapshotPluginId
+    // TODO remove synchronized after https://issues.scala-lang.org/browse/SI-7943
+    snapshotPlugins.synchronized(snapshotPlugins.getOrElseUpdate(configPath, createPlugin(configPath)(snapshotDispatchSelector)))
   }
 
   /**
    * Returns a journal for a persistent actor identified by `persistenceId`.
+   * When empty, looks in `akka.persistence.journal.plugin` to find journal configuration entry path.
+   * When configured, uses `journalPluginId` declared by the persistent actor, which must be live at the time of invocation.
    */
   def journalFor(persistenceId: String): ActorRef = {
```

Comment:
I think we should add `journalPluginId: String` to this method instead of the `register`, and thereby we don't need the `actorPlugins` map.